Offer to install helm into venv

mplsgrant opened the pull request:
This offers to install helm into the user's venv directory. Fixes #608

I overloaded the `is_helm_installed` function a bit, but I figured that even if we had to do this for every dependency it would remain manageable.

mplsgrant committed:
add constants related to the helm binary

mplsgrant committed:
`setup` shoehorn in an option to download helm

pinheadmz commented on src/warnet/constants.py:
```diff
@@ -98,3 +98,8 @@
 ]
+
+# Helm binary
+HELM_LATEST_URL = "https://get.helm.sh/helm-latest-version"
```
I think we should pin a version and also commit to the expected checksums:
https://github.com/helm/helm/releases

mplsgrant replied:
I decided to bless v3.16.1 and always pull it: https://github.com/bitcoin-dev-project/warnet/pull/610/commits/07104ee988914e568655dde1e324f752ca6887f8

pinheadmz commented:
` ⭐️ Helm is satisfied: /Users/matthewzipkin/Desktop/work/warnet/.venv/bin/helm`

This is fucking dope dude, great work

willcl-ark commented:
Yeah this is neat!

Perhaps it would help us to define the install methods we will offer/support (and stick to it!). Something like:

1. Manual: we list the packages you need and link to their install docs. Install them how you wish.
2. Semi-automatic: we offer to fetch bins directly into the local .venv (like here!) for all possible requirements.
3. Fully automatic: a docker image, something like that proposed in #615

Each of the methods should cover **all** required binaries. So for example I'd propose to follow this PR here with a similar `kubectl` and possibley even `k9s` install into venv.

If a user can't/won't install using one of these three ways, then they can't use the tool.

This would hopefully stop this fiasco re-ocurring in the future (or at least, we wouldn't have to worry about it).

mplsgrant committed:
`project`: add a better arch_map

mplsgrant committed:
`constants`: add blessed helm metadata

mplsgrant committed:
`project`: get blessed helm version

Get the blessed version instead of the latest version.

willcl-ark commented:
> Reviewers, this pull request conflicts with the following ones:
> 
>     * [#614](https://github.com/bitcoin-dev-project/warnet/pull/614) (swap out kubectl by mplsgrant)

@mplsgrant do you have a preference between which of these goes first? It looks like this one is perhaps more ready?

pinheadmz commented:
Install looks good but i got this error trying to deploy a network with logging. Not sure if this is an issue with the pinned version of Helm or something wrong with the loki chart in the repo?

warnet error:

`Exception: Error: UPGRADE FAILED: "loki" has no deployed releases`

So i tried to install using helm directly instead of through warnet...


```
$ helm upgrade --install --namespace warnet-logging --create-namespace --values ./resources/manifests/loki_values.yaml loki grafana/loki --version 5.47.2 --dry-run --debug
history.go:56: 2024-09-23 09:43:19.006541243 -0400 EDT m=+0.091649742 [debug] getting history for release loki
upgrade.go:164: 2024-09-23 09:43:19.874007994 -0400 EDT m=+0.959116488 [debug] preparing upgrade for loki
Error: UPGRADE FAILED: "loki" has no deployed releases
helm.go:86: 2024-09-23 09:43:20.456117492 -0400 EDT m=+1.541225945 [debug] "loki" has no deployed releases
UPGRADE FAILED
main.newUpgradeCmd.func2
        helm.sh/helm/v3/cmd/helm/upgrade.go:243
github.com/spf13/cobra.(*Command).execute
        github.com/spf13/cobra@v1.8.1/command.go:985
github.com/spf13/cobra.(*Command).ExecuteC
        github.com/spf13/cobra@v1.8.1/command.go:1117
github.com/spf13/cobra.(*Command).Execute
        github.com/spf13/cobra@v1.8.1/command.go:1041
main.main
        helm.sh/helm/v3/cmd/helm/helm.go:85
runtime.main
        runtime/proc.go:271
runtime.goexit
        runtime/asm_amd64.s:1695
```

mplsgrant committed:
`project`: fix arch<->uname mapping

mplsgrant commented:
> Install looks good but i got this error trying to deploy a network with logging. Not sure if this is an issue with the pinned version of Helm or something wrong with the loki chart in the repo?

@pinheadmz I noticed an issue with the architecture mapping. I'm not sure if that had something to do with the error you encountered, but I fixed that and it might be worth trying again.

mplsgrant commented:
> do you have a preference between which of these goes first? It looks like this one is perhaps more ready?

@willcl-ark I don't have a preference about which goes first. I'm fine to rebase either one.

pinheadmz commented:
ACK  ec33acea9a40894fbd792bce6a71241132fa2c76
seems to work fine now!

mplsgrant committed:
remove dead code

We no longer need the logic for the latest version of helm.
Also we don't need the NeedsHelm option in ToolStatus.

mplsgrant commented:
Just wanted to remove a bit of dead code before merging.

m3dwards commented:
On Arm Mac I got: `No Helm binary candidate for arch: None`

Also the question on setup is truncated: 

```shell
[?] Would you like to use Warnet's downloader to install Helm into your virtual environme...:
```

pinheadmz commented:
Can confirm arm macos issue, but the message printed fine for me

![Screenshot 2024-09-27 at 10 08 10 AM](https://github.com/user-attachments/assets/cb6c03ff-35f7-42a8-a68f-0751ee7fe108)

m3dwards commented:
The truncation must be because I run narrow terminals. Perhaps the question could be a bit shorter? Assuming ~~click~~ inquirer doesn't have a wrap text option.

mplsgrant commented:
> On Arm Mac I got: `No Helm binary candidate for arch: None`

@m3dwards What happens when you run this in a python terminal?

````python
import os
os.uname().machine
````

pinheadmz commented:
```
>>> import os
>>> os.uname()
posix.uname_result(sysname='Darwin', nodename=' ', release='23.5.0', version='Darwin Kernel Version 23.5.0: Wed May  1 20:16:51 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T8103', machine='arm64')
>>> os.uname().machine
'arm64'

```

mplsgrant committed:
shorten user prompt

mplsgrant committed:
support "arm64"

mplsgrant committed:
update missing arch error message

Show user their arch instead of showing them "None"

m3dwards commented:
> arm64

![](https://media.tenor.com/xjvb3NSHtzQAAAAM/same-ditto.gif)

m3dwards commented:
Works! Great PR!

pinheadmz commented:
post merge ACK,
tested on macos arm64 and debian x86_64
